### PR TITLE
Group location support

### PR DESCRIPTION
**Pull request by aliciaxw**
- `Group` has two new fields, `location: { lat: ?number, long: ?number }` and `isLocationRestricted: boolean`
- new endpoint `/sessions/:id/location/restriction/` that updates a group's location restriction
- `/start/session/` initializes the group location with the admin's location (lat/long can be null)
- `/join/session/` will update the group's location if the user is an admin, or possibly throw an error if the user is a member and is not within 300 meters of the group's location

### CODE REVIEW

File: src/repos/GroupsRepo.js [hunk 1]
```diff
@@ -125,6 +131,72 @@ const updateGroupByID = async (id: number, name: string):
   }
 };
 
+/**
+ * Update the location coordinates of a group
+ * @param {*} id - ID of group to update
+ * @param {*} coords - New coordinates of group
+ * @return {?Group} Updated group
+ */
+const updateGroupLocation = async (id: number, coords: Coord): Promise<?Group> => {
```
Comment (yuna878) [inline src/repos/GroupsRepo.js]: I believe there is a method to update group already? Idk if it's better to have a separate method for updating juts location or better to combine the two

File: src/repos/GroupsRepo.js [hunk 2]
```diff
@@ -125,6 +131,72 @@ const updateGroupByID = async (id: number, name: string):
...
+};
+
+/**
+ * Update a group's location restriction
+ * @function
+ * @param {number} id - ID of group to update
+ * @param {boolean} isRestricted - If joining a group is restricted by location
+ * @return {?Group} Updated group
+ */
+const updateLocationRestriction = async (id: number, isRestricted: boolean):
```
Comment (yuna878) [inline src/repos/GroupsRepo.js]: Same here with updating

File: src/routers/v2/JoinGroupRouter.js [hunk 3]
```diff
@@ -34,19 +35,41 @@ class JoinGroupRouter extends AppDevRouter<APIGroup> {
...
+      if (updatedGroup) group = updatedGroup;
+    }
+
+    // handle location restriction constraints for members
+    if (isMember && group.isLocationRestricted) {
+      if (!location || !location.lat || !location.long) {
+        throw LogUtils.logErr('User with no location tried to join a location restricted group');
+      }
+
+      if (!group.location.lat || !group.location.long) {
```
Comment (yuna878) [inline src/routers/v2/JoinGroupRouter.js]: Could there be an edge case where admin sets `isLocationRestricted = true` but location might be null (e.g. admin turned on setting for location restriction but didn't allow location sharing)? I don't think its intuitive to update the boolean setting if data doesn't exists. Maybe its safer to check for `isLocatedRestricted`

File: src/utils/Lib.js [hunk 4]
```diff
@@ -118,9 +120,29 @@ async function updateSession(req: Request, res: Response, next: NextFunction) {
...
+ */
+const isWithin300m = (userCoords: Coord, groupCoords: Coord): boolean => {
+  const lat2 = userCoords.lat;
+  const lat1 = groupCoords.lat;
+  const long2 = userCoords.long;
+  const long1 = groupCoords.long;
+  if (lat2 && lat1 && long2 && long1) {
+    const p = 0.017453292519943295; // pi / 180
+    const c = Math.cos;
+    const a = 0.5 - c((lat2 - lat1) * p) / 2 + c(lat1 * p) * c(lat2 * p) * (1 - c((long2 - long1) * p)) / 2;
```
Comment (yuna878) [inline src/utils/Lib.js]: More descriptive variable name or short comment on what this line is doing would be helpful

Comment (aliciaxw) [reply src/utils/Lib.js]: to be honest i'm not entirely sure what this does, i grabbed the formula from [wikipedia](https://en.wikipedia.org/wiki/Haversine_formula) lol

File: src/repos/GroupsRepo.js (hunk 1 above)
Comment (aliciaxw) [reply src/repos/GroupsRepo.js]: addressed in push!

File: src/routers/v2/Groups/UpdateLocationRestrictionRouter.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
+// @flow
```
Comment (yuna878) [inline src/routers/v2/Groups/UpdateLocationRestrictionRouter.js]: Can we combine location update into `UpdateGroupRouter.js`?

File: src/utils/Lib.js (hunk 4 above)
Comment (yuna878) [reply src/utils/Lib.js]: lol then can you change the comment in line 136 to something like `distance between userCoords and groupCoords in meters`

File: src/utils/Lib.js [hunk 6]
```diff
@@ -118,9 +120,29 @@ async function updateSession(req: Request, res: Response, next: NextFunction) {
...
+ * Checks if the new location is within 300 meters of the
+ * old location using Haversine's formula
+ */
+const isWithin300m = (userCoords: Coord, groupCoords: Coord): boolean => {
+  const lat2 = userCoords.lat;
+  const lat1 = groupCoords.lat;
+  const long2 = userCoords.long;
+  const long1 = groupCoords.long;
+  if (lat2 && lat1 && long2 && long1) {
+    const p = 0.017453292519943295; // pi / 180
```
Comment (meganle) [inline src/utils/Lib.js]: I think you can do `Math.PI / 180`

File: src/routers/v2/Groups/UpdateLocationRestrictionRouter.js (hunk 5 above)
Comment (meganle) [reply src/routers/v2/Groups/UpdateLocationRestrictionRouter.js]: If this is just updating the `isRestricted` field than it should be combined. What about updating the specific coordinates?

File: src/utils/Lib.js (hunk 4 above)
Comment (meganle) [reply src/utils/Lib.js]: +1 to yuna's comment

File: src/routers/v2/Groups/UpdateLocationRestrictionRouter.js (hunk 5 above)
Comment (yuna878) [reply src/routers/v2/Groups/UpdateLocationRestrictionRouter.js]: I think specific coordinates are updated in `JoinGroupRouter.js` when admin joins the group right now?

File: src/repos/GroupsRepo.js [hunk 7]
```diff
@@ -101,27 +107,49 @@ const deleteGroupByID = async (id: number) => {
...
+      .leftJoinAndSelect('groups.admins', 'admins')
+      .leftJoinAndSelect('groups.members', 'members')
+      .leftJoinAndSelect('groups.polls', 'polls')
+      .leftJoinAndSelect('groups.questions', 'questions')
+      .where('groups.id = :groupID', { groupID: id })
+      .getOne();
+
+    if (name) group.name = name;
+    if (location && location.lat && location.long) group.location = location;
+    if (isRestricted !== null || isRestricted !== undefined) group.isLocationRestricted = isRestricted;
```
Comment (meganle) [inline src/repos/GroupsRepo.js]: You can just do `if (isRestricted)` i think

File: src/routers/v2/Groups/UpdateGroupRouter.js [hunk 8]
```diff
@@ -17,11 +17,13 @@ class UpdateGroupRouter extends AppDevRouter<APIGroup> {
   }
 
   async content(req: Request) {
-    const { name } = req.body;
+    const { isRestricted, name } = req.body;
     const groupID = req.params.id;
     const { user } = req;
 
-    if (!name) throw LogUtils.logErr('No fields specified to update');
+    if (!name && (isRestricted === null || isRestricted === undefined)) {
```
Comment (meganle) [inline src/routers/v2/Groups/UpdateGroupRouter.js]: can you just do `if (!name && !isRestricted)`

File: src/utils/Lib.js [hunk 9]
```diff
@@ -5,8 +5,10 @@ import {
   Request,
 } from 'express';
 import profanity from 'profanity-util';
+import type { Coord } from '../models/Group';
```
Comment (meganle) [inline src/utils/Lib.js]: i would put the import of types separate from these imports like in `JoinGroupRouter`

File: test/routes/api.group.test.js [hunk 10]
```diff
@@ -191,6 +214,7 @@ test('delete group with invalid adminToken', async () => {
 
 test('delete group', async () => {
   const result = await request(del(`/sessions/${group.id}`, adminToken));
+  console.log(result);
```
Comment (meganle) [inline test/routes/api.group.test.js]: delete console log

File: src/repos/GroupsRepo.js [hunk 11]
```diff
@@ -3,6 +3,7 @@ import { getConnectionManager, Repository } from 'typeorm';
 import LogUtils from '../utils/LogUtils';
 import Poll from '../models/Poll';
 import Question from '../models/Question';
+import type { Coord } from '../models/Group';
```
Comment (meganle) [inline src/repos/GroupsRepo.js]: you should keep import types separate from normal imports (see similar comment below)

File: src/repos/GroupsRepo.js (hunk 7 above)
Comment (aliciaxw) [reply src/repos/GroupsRepo.js]: `isRestricted` is a boolean tho so wouldn't `if (isRestricted)` not be satisfied if `isRestricted === false`

File: src/routers/v2/JoinGroupRouter.js (hunk 3 above)
Comment (aliciaxw) [reply src/routers/v2/JoinGroupRouter.js]: resolved, we just throw an error instead of updating the setting

File: src/utils/Lib.js (hunk 4 above)
Comment (aliciaxw) [reply src/utils/Lib.js]: changed

**Pull request merged**